### ISSUE TITLE
GCC High - Endpoint Used Returning no data

### ISSUE THREAD

**Issue by cyanitol** [counteractive/o365beat#25]
I don't see an option in the settings to change the endpoint API URLs.  When attempting to use this for GCC high environments it will not work and returns empty data due to the wrong endpoint being targeted for the API.

**Endpoint Mapping information can be found here:**
https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/active-directory/fundamentals/whats-new-microsoft-365-government.md

https://docs.microsoft.com/en-us/office365/enterprise/office-365-u-s-government-gcc-high-endpoints

**Comment by chris-counteractive** [comment]
Great idea, this should be a straightforward enhancement.  We don't have a GCC high instance to test on and the docs don't cover whether there are any differences when using those endpoints - would you be able to let me know if it works once I get you a build that lets you customize those endpoints?  Should be soon.

The relevant code is [here](https://github.com/counteractive/o365beat/blob/master/beater/o365beat.go#L63-L64), I'll look to add config variables that default to the "normal" (read, commercial) endpoints, but can be updated for other environments like so:

```yaml
o365beat:
  login_url: login.microsoftonline.us  # default is https://login.microsoftonline.com/
  resource_url: manage.office365.us    # default is https://manage.office.com/
  # rest of your config ...
```

**Comment by cyanitol** [comment]
Awesome.  I made some modifications to the endpoints for testing the other day, and it has been working excellent.  Diff is below:

`diff --git a/beater/o365beat.go b/beater/o365beat.go
index f7cbde5..b1f862f 100644
--- a/beater/o365beat.go
+++ b/beater/o365beat.go
@@ -60,8 +60,8 @@ func New(b *beat.Beat, cfg *common.Config) (beat.Beater, error) {
        }
 
        // using url.Parse seems like overkill
-       loginURL := "https://login.microsoftonline.com/"
-       resourceURL := "https://manage.office.com/"
+       loginURL := "https://login.microsoftonline.us/"
+       resourceURL := "https://manage.office365.us/"
        au := loginURL + c.TenantDomain + "/oauth2/token?api-version=1.0"
        api := resourceURL + "api/v1.0/" + c.DirectoryID + "/activity/feed/"
        cl := &http.Client{Timeout: c.APITimeout}
@@ -89,7 +89,7 @@ func (bt *O365beat) apiRequest(verb, urlStr string, body, query, headers map[str
                logp.Error(err)
                return nil, err
        }
-       reqQuery := req.URL.Query() // keep querystring values from urlStr
+       reqQuery := req.URL.Query()                                // keep query:...skipping...
diff --git a/beater/o365beat.go b/beater/o365beat.go
index f7cbde5..b1f862f 100644
--- a/beater/o365beat.go
+++ b/beater/o365beat.go
@@ -60,8 +60,8 @@ func New(b *beat.Beat, cfg *common.Config) (beat.Beater, error) {
        }
 
        // using url.Parse seems like overkill
-       loginURL := "https://login.microsoftonline.com/"
-       resourceURL := "https://manage.office.com/"
+       loginURL := "https://login.microsoftonline.us/"
+       resourceURL := "https://manage.office365.us/"
        au := loginURL + c.TenantDomain + "/oauth2/token?api-version=1.0"
        api := resourceURL + "api/v1.0/" + c.DirectoryID + "/activity/feed/"
        cl := &http.Client{Timeout: c.APITimeout}
@@ -89,7 +89,7 @@ func (bt *O365beat) apiRequest(verb, urlStr string, body, query, headers map[str
                logp.Error(err)
                return nil, err
        }
-       reqQuery := req.URL.Query() // keep querystring values from urlStr
+       reqQuery := req.URL.Query()                                // keep querystring values from urlStr
        reqQuery.Set("PublisherIdentifier", bt.config.DirectoryID) // to prevent throttling
        for k, v := range query {
                reqQuery.Set(k, v)
@@ -130,7 +130,7 @@ func (bt *O365beat) authenticate() error {
        logp.Info("authenticating via %s", bt.authURL)
        reqBody := url.Values{}
        reqBody.Set("grant_type", "client_credentials")
-       reqBody.Set("resource", "https://manage.office.com")
+       reqBody.Set("resource", "https://manage.office365.us")
        reqBody.Set("client_id", bt.config.ClientID)
        reqBody.Set("client_secret", bt.config.ClientSecret)
        req, err := http.NewRequest("POST", bt.authURL, strings.NewReader(reqBody.Encode()))
@@ -180,7 +180,7 @@ func (bt *O365beat) subscribe(contentType string) (common.MapStr, error) {
        logp.Info("note that new subscriptions can take up to 12 hours to produce data")
        logp.Debug("api", "subscribing to %v at %s", contentType, bt.apiRootURL+"subscriptions/start")
        query := map[string]string{
-               "contentType":         contentType,
+               "contentType": contentType,
        }
        res, err := bt.apiRequest("POST", bt.apiRootURL+"subscriptions/start", nil, query, nil)
        if err != nil {
@@ -259,9 +259,9 @@ func (bt *O365beat) listAvailableContent(contentType string, start, end time.Tim
 
        dateFmt := "2006-01-02T15:04:05" // API needs UTC in this format (no "Z" suffix)
        query := map[string]string{
-               "contentType":         contentType,
-               "startTime":           start.UTC().Format(dateFmt),
-               "endTime":             end.UTC().Format(dateFmt),
+               "contentType": contentType,
+               "startTime":   start.UTC().Format(dateFmt),
+               "endTime":     end.UTC().Format(dateFmt),
        }
        res, err := bt.apiRequest("GET", bt.apiRootURL+"subscriptions/content", nil, query, nil)
        if err != nil {
~
~
~
`

**Comment by cyanitol** [comment]
Awesome. I made some modifications to the endpoints for testing the other day, and it has been working excellent. Diff attached.
[diff.txt](https://github.com/counteractive/o365beat/files/3947271/diff.txt)

**Comment by chris-counteractive** [comment]
Gotta love open source, thanks for testing that, and thanks for reaching out so others can benefit from it.  I'll work a configurable option into a new build soon, thanks again.

**Comment by chris-counteractive** [comment]
The beat pulls the login url and resource urls from the config file as of https://github.com/counteractive/o365beat/commit/dcdf24e4a3d4ff9fac69c3abc1c7a67e33847752, I'll update the documentation and wrap it all into the next release, and that should do it for this one.  Thanks again for the issue.

**Comment by chris-counteractive** [comment]
Released in [v1.5.0](https://github.com/counteractive/o365beat/releases/tag/v1.5.0).
